### PR TITLE
feature/integration

### PR DESCRIPTION
**Pull request by deebloo**
this adds a demo app to the repo. We should discuss what exactly a demo like this would look like, this is just a starting point. would address #11 Will use this as a base to start adding more integration tests

### CODE REVIEW

File: demo/app/app.component.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+    <ul>
+      <li *ngFor="let todo of todos$ | async; let i = index">
+        {{todo}} <button (click)="removeTodo(i)">remove</button>
+      </li>
+    </ul>
+  `
+})
+export class AppComponent {
+  todos$ = this.ngxs.select(state => state.todos);
```
Comment (amcdnl) [inline demo/app/app.component.ts]: This should use the `@Select` decorator

File: demo/app/app.component.ts [hunk 2]
```diff
@@ -18,7 +19,7 @@ import { AddTodo, RemoveTodo } from './stores/todo.store';
   `
 })
 export class AppComponent {
-  todos$ = this.ngxs.select(state => state.todos);
+  @Select('todos') todos$: Observable<string[]>;
```
[the remark this answers is not in the document]
Comment (deebloo) [reply demo/app/app.component.ts]: ooooooooooooooo.

[the remark this answers is not in the document]
Comment (SebastianM) [reply demo/app/app.component.ts]: Hmm... does this work with minifiers? I mean this: `@Select() todos$`

[the remark this answers is not in the document]
Comment (amcdnl) [reply demo/app/app.component.ts]: Well crap, your probably right. but only on aggressive ones.

[the remark this answers is not in the document]
Comment (kylecordes) [reply demo/app/app.component.ts]: The comment about minifiers points to a challenge with this type of syntax. By moving any aspect of the code (in this case, an identifier) into a string, it becomes very challenging to process the code:

* Aggressive minification
* IDE support for renaming things
* Human search and replace, in which you must spot identifiers both inside and outside strings

At the risk of being boring, how about following the approach other libraries have followed, favoring a kind of "selector library" to help concisely create functions that can be reasoned about by the type system, composed, etc.?

[the remark this answers is not in the document]
Comment (amcdnl) [reply demo/app/app.component.ts]: Ya, I agree. Personally I find the `Select(state => state.foo.bar)` to be verbose which is why I made the option for strings but with that you loose type safety. I think the contents of the string should be safe though?

[the remark this answers is not in the document]
Comment (deebloo) [reply demo/app/app.component.ts]: I agree with @kylecordes. I also think Ngxs should be generic so you can define the app state

**Comment by jotatoledo** [comment]
Could this demo app include:

- Store composition: create another store class and showcase what are the options to combine it with the todostore into 1.
- Selectors: showcase all the available options create em through the decorator.
- Actions: same as selectors
- Dispatch + subscribe
- The use of NgxsModule.forFeature() method
- Lazy loaded stores?

**Comment by deebloo** [comment]
Im going to work on this later today. Ill add a checklist to see what all we want to add. This is also meant to act as an integration test so we need to account for as much as possible

**Comment by amcdnl** [comment]
Let me know when this is ready :)

**Comment by deebloo** [comment]
@amcdnl hey. I think this is good to go. It lays the groundwork for integration test in addition to the libraries own unit tests. The list of things we WANT to have covered here I can do in a separate PR

**Comment by deebloo** [comment]
cc @jotatoledo ☝️

**Comment by deebloo** [comment]
@amcdnl also I would suggest squashing and merging this one

**Comment by amcdnl** [comment]
Merged

**Pull request merged**